cli/src/main/java/com/devonfw/tools/ide/context/GitUtils.java

```diff
@@ -0,0 +1,201 @@
+package com.devonfw.tools.ide.context;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.List;
+
+import com.devonfw.tools.ide.cli.CliException;
+import com.devonfw.tools.ide.process.ProcessContext;
+import com.devonfw.tools.ide.process.ProcessErrorHandling;
+import com.devonfw.tools.ide.process.ProcessResult;
+
+/**
+ * Provides utilities for git.
+ */
+public class GitUtils {
+  private final IdeContext context;
+
+  private ProcessContext processContext;
+
+  private final Path targetRepository;
+
+  private final String remoteName;
+
+  private final String branchName;
+
+  /**
+   * @param context the {@link IdeContext context}.
+   * @param targetRepository the target repository Path.
+   * @param remoteName the remote server name.
+   * @param branchName the name of the branch.
+   */
+  public GitUtils(IdeContext context, Path targetRepository, String remoteName, String branchName) {
+
+    this.context = context;
+    this.targetRepository = targetRepository;
+    this.remoteName = remoteName;
+    this.branchName = branchName;
+
+  }
+
+  /**
+   * Runs a git pull or a git clone.
+   *
+   * @param force boolean true enforces a git hard reset and cleanup of added files.
+   * @param gitRepoUrl String of repository URL.
+   */
+  protected void runGitPullOrClone(boolean force, String gitRepoUrl) {
+
+    this.processContext = this.context.newProcess().directory(this.targetRepository).executable("git")
+        .withEnvVar("GIT_TERMINAL_PROMPT", "0");
```

Comment:
here ProcessContext is created but it is also used in other methods. All of these methods are protected (none is public).
If one of the other methods is used before this one, it will fail with a NullPointerException.
I would consider creating a private method to access `processContext` that creates it lazily and all methods access it via the private method. It is also possible to not store it as member variable and always create it per method, but reuse can make sense...

Reply:
In my new test scenarios I actually ran into this problem already.
I agree that we need to re-initialize this before each git command.